docker/src/main/java/brooklyn/location/docker/DockerContainerLocation.java

```diff
@@ -168,20 +171,57 @@ public HostAndPort getSocketEndpointFor(Cidr accessor, int privatePort) {
 
     @Override
     public int execScript(Map<String,?> props, String summaryForLogging, List<String> commands, Map<String,?> env) {
-        Iterable<String> filtered = Iterables.filter(commands, DockerCallbacks.FILTER);
-        for (String commandString : filtered) {
-            parseDockerCallback(commandString);
+        if(getOwner().config().get(DockerContainer.DOCKER_USE_EXEC)) {
+            Map<String,?> nonPortProps = Maps.filterKeys(props, Predicates.not(Predicates.containsPattern("port")));
+            SshMachineLocation host = getOwner().getDockerHost().getDynamicLocation().getMachine();
+            return host.execCommands(nonPortProps, summaryForLogging, getExecScript(commands, env));
+        } else {
+            Iterable<String> filtered = Iterables.filter(commands, DockerCallbacks.FILTER);
+            for (String commandString : filtered) {
+                parseDockerCallback(commandString);
+            }
+            return super.execScript(props, summaryForLogging, commands, env);
         }
-        return super.execScript(props, summaryForLogging, commands, env);
     }
 
     @Override
     public int execCommands(Map<String,?> props, String summaryForLogging, List<String> commands, Map<String,?> env) {
-        Iterable<String> filtered = Iterables.filter(commands, DockerCallbacks.FILTER);
-        for (String commandString : filtered) {
-            parseDockerCallback(commandString);
+        if(getOwner().config().get(DockerContainer.DOCKER_USE_EXEC)) {
+
+            SshMachineLocation host = getOwner().getDockerHost().getDynamicLocation().getMachine();
+            Map<String,?> nonPortProps = Maps.filterKeys(props, Predicates.not(Predicates.containsPattern("port")));
+            return host.execCommands(nonPortProps, summaryForLogging, getExecCommands(commands, env));
+        } else {
+            Iterable<String> filtered = Iterables.filter(commands, DockerCallbacks.FILTER);
+            for (String commandString : filtered) {
+                parseDockerCallback(commandString);
+            }
+            return super.execCommands(props, summaryForLogging, commands, env);
         }
-        return super.execCommands(props, summaryForLogging, commands, env);
+    }
+
+    private List<String> getExecScript(List<String> commands, Map<String,?> env) {
+        String prefix = "docker exec " + dockerContainer.getContainerId() + " '";
+        return Collections.singletonList(prefix + Joiner.on(';').join(
```

Comment:
use `ImmutableList` to create singleton and build script command from `StringBuilder` instead